### ISSUE TITLE
too_many_buckets_exception is thrown regardless of size/shard_size setting in query

### ISSUE THREAD

**Issue by nescalera** [opendatasoft/elasticsearch-aggregation-pathhierarchy#17]
Hello there,

We encountered an issue when aggregating data that creates more buckets than the specified by the cluster's **max_buckets** setting.  I attempted to restrict the number of buckets that will be returned by using **size** and **shard_size** in the aggregation without much luck. Here's an easy way to reproduce the issue:

**Plugin versions**: 7.4.2 and 7.6.0

```
##################################
##### Index settings/mappings
##################################
PUT /filesystem
{
  "settings": {
    "number_of_shards": 1,
    "number_of_replicas": 1
  },
  "mappings": {
    "properties":{
      "file": {
        "properties": {
         "path": {
            "type": "text",
            "fields": {
              "keyword": {
                "type": "keyword"
              }
            }
          }
        }
      }
    }
  }
}

##################################
##### Cluster Settings
##################################
PUT _cluster/settings
{
  "transient": {
    "search.max_buckets": 3
  }
}

##################################
##### Documents
##################################
PUT /filesystem/_doc/1
{
  "path": "/My documents/Spreadsheets/Budget_2013.xls"
}

PUT /filesystem/_doc/2
{
  "path": "/My documents/Spreadsheets/Budget_2014.xls"
}

PUT /filesystem/_doc/3
{
  "path": "/My documents/Spreadsheets/Budget_2015.xls"
}

PUT /filesystem/_doc/4
{
  "path": "/My documents/Test.txt"
}

```
This search query is using `path_hierarchy` for the aggregations **with** `size` and `shard_size`:
```
GET /filesystem/_search
{
  "size": 0,
  "aggs": {
    "tree": {
      "path_hierarchy": {
        "field": "path.keyword",
        "separator": "/",
        "size": 3,
        "shard_size": 3
        
      }
    }
  }
}
```

Result:
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "too_many_buckets_exception",
        "reason" : "Trying to create too many buckets. Must be less than or equal to: [3] but was [4]. This limit can be set by changing the [search.max_buckets] cluster level setting.",
        "max_buckets" : 3
      }
    ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query",
    "grouped" : true,
    "failed_shards" : [
      {
        "shard" : 0,
        "index" : "filesystem",
        "node" : "fffHTB1SQmmsOii8-cO4ew",
        "reason" : {
          "type" : "too_many_buckets_exception",
          "reason" : "Trying to create too many buckets. Must be less than or equal to: [3] but was [4]. This limit can be set by changing the [search.max_buckets] cluster level setting.",
          "max_buckets" : 3
        }
      }
    ]
  },
  "status" : 503
}
```

When a similar query is executed using a _terms aggregation_ instead, the error is not thrown as long as the `size` and `shard_size` are provided:

Terms aggregation **WITHOUT** specifying bucket/shard size
```
GET /filesystem/_search?pretty
{
  "size": 0,
  "query": {
    "bool": {
      "must": [
        {
          "match_all": {}
        }
      ]
    }
  },
  "aggs": {
      "tree": {
        "terms": {
          "field": "path.keyword"
        }
      }
  }
}
```

Result:
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "too_many_buckets_exception",
        "reason" : "Trying to create too many buckets. Must be less than or equal to: [3] but was [4]. This limit can be set by changing the [search.max_buckets] cluster level setting.",
        "max_buckets" : 3
      }
    ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query",
    "grouped" : true,
    "failed_shards" : [
      {
        "shard" : 0,
        "index" : "filesystem",
        "node" : "fffHTB1SQmmsOii8-cO4ew",
        "reason" : {
          "type" : "too_many_buckets_exception",
          "reason" : "Trying to create too many buckets. Must be less than or equal to: [3] but was [4]. This limit can be set by changing the [search.max_buckets] cluster level setting.",
          "max_buckets" : 3
        }
      }
    ]
  },
  "status" : 503
}
```
Terms aggregation **WITH** bucket/shard size
```
GET /filesystem/_search?pretty
{
  "size": 0,
  "query": {
    "bool": {
      "must": [
        {
          "match_all": {}
        }
      ]
    }
  },
  "aggs": {
      "tree": {
        "terms": {
          "field": "path.keyword",
          "size": 3,
          "shard_size": 3
        }
      }
  }
}
```

Result:
```
{
  "took" : 1,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 4,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "tree" : {
      "doc_count_error_upper_bound" : 0,
      "sum_other_doc_count" : 1,
      "buckets" : [
        {
          "key" : "/My documents/Spreadsheets/Budget_2013.xls",
          "doc_count" : 1
        },
        {
          "key" : "/My documents/Spreadsheets/Budget_2014.xls",
          "doc_count" : 1
        },
        {
          "key" : "/My documents/Spreadsheets/Budget_2015.xls",
          "doc_count" : 1
        }
      ]
    }
  }
}
```

Am I missing something in my configuration?  I would love to hear your feedback!

Thanks in advance,
-Natalia

**Comment by rkouye** [comment]
Hello @nescalera

It is indeed a bug, and we are working on a [fix](https://github.com/rkouye/elasticsearch-aggregation-pathhierarchy/pull/1). A PR should come soon but feel free to test it.

**Comment by nescalera** [comment]
Great! Thanks for the update @rkouye

**Comment by nescalera** [comment]
@rkouye Thank you for fixing this issue.  I tried the changes in a test environment and they worked like a charm.  Any idea when the next release will be available?

**Comment by rkouye** [comment]
Thanks, @nescalera. Which version did you test? The one on ES 7.6 https://github.com/rkouye/elasticsearch-aggregation-pathhierarchy/pull/1 or the one on ES 7.9 https://github.com/opendatasoft/elasticsearch-aggregation-pathhierarchy/pull/19?

FYI, we are not planning to release soon as we need more tests before we could confidently deploy in prod.

**Comment by nescalera** [comment]
@rkouye I tested with https://github.com/opendatasoft/elasticsearch-aggregation-pathhierarchy/pull/19

**Comment by nescalera** [comment]
Hi @rkouye, quick question, was this issue addressed in v7.10.2.0?

**Comment by rkouye** [comment]
Hi @nescalera, I will let @benj-moreau confirm that

**Comment by benj-moreau** [comment]
Hi @nescalera, 

I can confirm you that the fix is implemented since the release of v7.10.2.0 I hope it will meets all your needs.
If not, don't hesitate to re-open an issue.

Thank you again for reporting this bug.
